### ISSUE TITLE
Post calls not working through CCFE

### ISSUE THREAD

**Issue by chetann** [linkedin/cruise-control-ui#10]
@nareshv Here is the scenario what I am seeing. 
Whenever I am on  `Kafka Cluster administration` tab and trying to execute any of the task such as "Preferred Leader Election" "Rebalance Cluster" "Add Brokers" "Remove Brokers" "Demote Brokers" etc, I am getting `Error: Network Error`.
Though when I inspect, my console shows me `Has been blocked by CORS policy`, I frankly feel this is something to do more then CORS policy. Because of two reasons
1. All http `GET` calls works fine. (if CORS issue then even this should have failed)
2. Secondly, what I see when I click on any of operations I want to do, as shown in the attached image the request method goes in as `OPTIONS` instead of `POST` (which UI should have used) and I see empty response even though the status shows as `200 OK`.
And when CCFE calls the OPTIONS method, it goes to this point in servlet where nothing is defined for OPTIONS method
https://github.com/linkedin/cruise-control/blob/master/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/KafkaCruiseControlServlet.java#L103

To cross verify, I took the URL generated by CCFE and call it using POSTMAN as POST method, I am able to get response here but not from CCFE. Screenshot attached for more reference.

![screen shot 2019-03-05 at 3 27 30 pm](https://user-images.githubusercontent.com/3926813/53801759-02efe100-3f66-11e9-8a99-ebebbd649355.png)

**Comment by chetann** [comment]
@nareshv you have anything for me over here. I am not able to test any activity for those POST calls.

**Comment by nareshv** [comment]
@chetann , i am in the process of writing the wiki for the CORS method. This should help unblock you.

**Comment by nareshv** [comment]
@chetann , CORS is bit advanced to get it working than the Reverse proxy method and needs more security around it.

Regarding your observations

1 CORS does preflight check using OPTIONS Method
2 The responses from CC Doesn't include 'Access-Control-Allow-Origin' header, which is why the app is showing `Network Error` even though the response is `200 OK`

I have created these wikis to get this up and running with CORS

1) https://github.com/linkedin/cruise-control-ui/wiki/CORS-Method
2) https://github.com/linkedin/cruise-control-ui/wiki/Multiple-Kafka-Clusters-(CORS-Method)

From my initial testing so far, looks like CC has gone through few code restructuring and the header is not sent for some end points. I will check with CC team on this.

**Comment by chetann** [comment]
Hi @nareshv I want to come again clear on the above query I have mentioned.
1. The CC GET calls like `cluster state` `cluster load` etc works fine.
![image](https://user-images.githubusercontent.com/3926813/54082829-44570680-4341-11e9-9040-51e789374525.png)
2. The CC POST calls such as rebalance cluster, delete broker etc fails and that is the screenshot I have attached above in query above.
![image](https://user-images.githubusercontent.com/3926813/54082859-a879ca80-4341-11e9-8b8e-1a7520b131d5.png)


If it was a CORS issue, I should have not got response even for CC GET calls.

And to add on more information, I have made the changes you have mentioned earlier itself.

`Make changes to CC config/cruisecontrol.properties file as shown below
# Whether CORS support is enabled for API or not
# (default) webserver.http.cors.enabled=false
webserver.http.cors.enabled=true

# Value for Access-Control-Allow-Origin
# (default) webserver.http.cors.origin=http://localhost:8080/
# (really brave users) webserver.http.cors.origin=*
webserver.http.cors.origin=http://cc-host.example.com/

# Value for Access-Control-Request-Method
webserver.http.cors.allowmethods=OPTIONS,GET,POST

# Headers that should be exposed to the Browser (Webapp)
# This is a special header that is used by the
# User Tasks subsystem and should be explicitly
# Enabled when CORS mode is used as part of the
# Admin Interface
webserver.http.cors.exposeheaders=User-Task-ID`

 
Also in the CC source code I did add these section in the servlet for it to make it work for all the methods defined i.e. doGET, doPOST and doOption. 
```
    if (_corsEnabled) {
      response.setHeader("Access-Control-Allow-Credentials",
              "true");
      response.setHeader("Access-Control-Allow-Origin",
                         _config.getString(KafkaCruiseControlConfig.WEBSERVER_HTTP_CORS_ORIGIN_CONFIG));
      // This is required only as part of pre-flight response
      response.setHeader("Access-Control-Request-Method",
                         _config.getString(KafkaCruiseControlConfig.WEBSERVER_HTTP_CORS_ALLOWMETHODS_CONFIG));
    }
```
Post these changes i.e. one in CC properties file and in the above servlet, I was able to get the proper responses.

But when I tested my UI for POST calls like rebalance, demote broker etc etc that is when I am facing CORS error. I wonder because if CORS error then I have handled it (even you have mentioned the same in your document) and it does not throw error for GET calls and throws error only during POST calls. I hope this clears my query in detail.

**Comment by nareshv** [comment]
@chetann , for the time being, can you try using this plugin(s) [for firefox](https://addons.mozilla.org/en-US/firefox/addon/cors-everywhere/) and [for chrome](https://chrome.google.com/webstore/detail/allow-cors-access-control/lhobafahddgcelffkeicbaginigeejlf?hl=en) and tell me if you are still getting into these errors ?

**Comment by chetann** [comment]
which plugin @nareshv

**Comment by nareshv** [comment]
@chetann , the url highlited ones are links to browser plugins (you might have to activate these once installed).

Also, i have written a simple reverse-proxy in node.js which takes the `config.csv` that contains CC REST API absolute urls. See if this[1]  works.

[1] https://gist.github.com/nareshv/85858d67020d4da64c2d9d0f8af20c92

**Comment by chetann** [comment]
@nareshv the plugin worked (CONDITIONALLY) for now.
1. If the plugin is switched on = Post call Works and GET calls Fails
2. If the plugin is switched off = Post call Fails and GET call Works.
To get everyone onboard this way in team is tough, is there any permanent solution that you can provide. Thanks.

**Comment by nareshv** [comment]
@chetann , thanks for the confirmation.

This is indeed a CC issue, which needs proper fix on the [server side](https://github.com/linkedin/cruise-control/issues/585) 

Until then you have these choices.

1. Run CCFE in reverseproxy setup.
2. Run CCFE in CORS Mode + use browser plugins to fix server side issue temporarily.

I'll am bringing this to the CC dev attention for proper fix on server side.

**Comment by chetann** [comment]
Sure, can you kindly also mention to CC team @nareshv that the configs for CORS mentioned in cruisecontrol properties file does not behave as expected. Like even if we set to TRUE it does not work until in servlet try { } section is added to handle CORS issue. Strange. Thanks.

**Comment by chetann** [comment]
@nareshv Is there any permanent solution provided on the CORS issue. You had mentioned for time being we can use the plugin but is there any permanent work around provided by CC team or you on this. Kindly update.

**Comment by chetann** [comment]
@nareshv any update for me over here please.

**Comment by nareshv** [comment]
@chetann , like i mentioned before, CC team will add support for the CORS related headers, until then you can use reverseproxy mode or browser plugin.

**Comment by nareshv** [comment]
https://github.com/linkedin/cruise-control/commit/9c7ad36017421b2a0265c0d1ab024730b7d4e9de onwards CORS support is there in cruise-control.

Config changes required on CC to use CORS is documented in https://github.com/linkedin/cruise-control-ui/wiki/CORS-Method

**Comment by nareshv** [comment]
please create new issue if any more help is required.
